segmentstore/storage/src/test/java/io/pravega/segmentstore/storage/chunklayer/ChunkedSegmentStorageTests.java

```diff
@@ -1505,6 +1502,191 @@ private void testTruncate(long maxChunkLength, long truncateAt, int chunksCountB
+     *
+     * @throws Exception Exception if any.
+     */
+    @Test
+    public void testReadWriteWithMultipleFailoversWithGarbage() throws Exception {
+        String testSegmentName = "foo";
+        TestContext testContext = getTestContext();
+
+        // Create
+        testContext.storageManager.create(testSegmentName, null).get();
+
+        // Write some data.
+        long writeAt = 0;
+        long epoch = 42;
+        SegmentHandle hWrite =  testContext.storageManager.openWrite(testSegmentName).get();
+        ArrayList<Long> lengths = new ArrayList<>();
+        for (int i = 1; i < 5; i++) {
+            // Create a new test context and initialize with new epoch.
+            testContext.storageManager.write(hWrite, writeAt, new ByteArrayInputStream(new byte[i]), i, null).join();
+            writeAt += i;
+            lengths.add((long) i);
+
+            // Read in same epoch.
+            checkDataRead(testSegmentName, testContext, 0, writeAt);
+
+            TestUtils.checkSegmentLayout(testContext.metadataStore, testSegmentName, Longs.toArray(lengths));
+
+            // Fork the context.
+            val oldTestCotext = testContext;
+            testContext = oldTestCotext.fork(epoch++);
+            TestUtils.checkSegmentLayout(testContext.metadataStore, testSegmentName, Longs.toArray(lengths));
+
+            // Make sure to open segment with new instance before writing garbage to old instance.
+            hWrite =  testContext.storageManager.openWrite(testSegmentName).get();
+
+            // Write some garbage
+            oldTestCotext.storageManager.write(hWrite, writeAt, new ByteArrayInputStream(new byte[10]), 10, null).join();
+
+            // Fence out old store.
+            boolean exceptionThrown = false;
+            oldTestCotext.metadataStore.markFenced();
+
+            AssertExtensions.assertFutureThrows("conact() allowed for invalid parameters",
+                    oldTestCotext.storageManager.write(hWrite, writeAt + 10, new ByteArrayInputStream(new byte[10]), 10, null),
+                    ex -> ex instanceof StorageNotPrimaryException);
+            // Read in new epoch.
+            checkDataRead(testSegmentName, testContext, 0, writeAt);
+        }
+
+        int total = 10;
+
+        // Create a new test context and initialize with new epoch.
+        testContext = testContext.fork(epoch++);
+
+        checkDataRead(testSegmentName, testContext, 0, total);
+    }
+
+    /**
+     * Test truncate, read and write with multiple failovers.
+     *
+     * @throws Exception Exception if any.
+     */
+    @Test
+    public void testTruncateWithMultipleFailoversWithGarbage() throws Exception {
+        String testSegmentName = "foo";
+        TestContext testContext = getTestContext();
+
+        // Create
+        testContext.storageManager.create(testSegmentName, null).get();
+
+        // Write some data.
+        long writeAt = 0;
+        long truncateAt = 0;
+        long epoch = 42;
```

Comment:
Can you make 42 and 10 class-level constants? They are used everywhere in these tests.

Reply:
Fixed for 42.
Number 10 is used in each test with different meaning in different tests. So using common test wide contsnat isn't much useful. so living it as it is.